src/main/java/com/mixpanel/android/mpmetrics/MixpanelPushNotification.java

```diff
@@ -126,52 +179,80 @@ protected PendingIntent createActionIntent(String uri) {
         );
     }
 
-    protected void maybeSetChannel() {
+    protected void maybeSetChannel(NotificationData data) {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
             NotificationManager mNotificationManager =
                     (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
 
-            String channelId = MPConfig.getInstance(context).getNotificationChannelId();
+            String channelId = data.channelId == null ? MPConfig.getInstance(context).getNotificationChannelId() : data.channelId;
             String channelName = MPConfig.getInstance(context).getNotificationChannelName();
-            int importance = MPConfig.getInstance(context).getNotificationChannelImportance();
 
-            NotificationChannel channel = new NotificationChannel(channelId, channelName, importance);
-            int notificationDefaults = MPConfig.getInstance(context).getNotificationDefaults();
+            NotificationChannel channel = new NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_DEFAULT);
+            mNotificationManager.createNotificationChannel(channel);
 
-            if (notificationDefaults == Notification.DEFAULT_VIBRATE || notificationDefaults == Notification.DEFAULT_ALL) {
-                channel.enableVibration(true);
-            }
+            builder.setChannelId(channelId);
+        }
+    }
 
-            if (notificationDefaults == Notification.DEFAULT_LIGHTS || notificationDefaults == Notification.DEFAULT_ALL) {
-                channel.enableLights(true);
-                channel.setLightColor(Color.WHITE);
-            }
+    protected void maybeSetNotificationBadge(NotificationData data) {
+        if (data.badgeCount > 0) {
+            builder.setNumber(data.badgeCount);
+        }
+    }
 
-            mNotificationManager.createNotificationChannel(channel);
+    protected void maybeSetTime(NotificationData data) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && data.timeString != null) {
+            Instant instant = Instant.parse( "2011-05-03T11:58:01Z" );
+            builder.setWhen(instant.toEpochMilli());
+        }
+    }
 
-            builder.setChannelId(channelId);
+    protected void maybeSetVisibility(NotificationData data) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            builder.setVisibility(data.visibility);
         }
     }
 
     protected Bitmap getBitmapFromResourceId(int resourceId) {
         return BitmapFactory.decodeResource(context.getResources(), resourceId);
     }
 
+    protected Bitmap getBitmapFromUrl(String url) {
+        ImageStore is = new ImageStore(context, "MixpanelPushNotification");
+        try {
+            Bitmap bm = is.getImage(url);
+            return bm;
```

Comment:
just return `is.getImage(url)` no need to create a local var

Reply:
I thought we wanted to return null instead of throwing the exception or does that happen anyways?